packages/atclient/src/atclient.c

```diff
@@ -227,13 +227,22 @@ void atclient_set_read_timeout(atclient *ctx, int timeout_ms) {
 static int atclient_start_atserver_connection(atclient *ctx, const char *secondaryhost, const int secondaryport) {
   int ret = 1; // error by default
 
+  // remove hooks to preserve them across resets
+  atclient_connection_hooks *conn_hooks = ctx->atserver_connection.hooks;
+  ctx->atserver_connection.hooks = NULL;
+
+  // clear the atserver connection
   atclient_connection_free(&(ctx->atserver_connection));
   ctx->atserver_connection_started = false;
   memset(&(ctx->atserver_connection), 0, sizeof(atclient_connection));
 
+  // (re) initialize the atserver connection
   atclient_connection_init(&(ctx->atserver_connection), ATCLIENT_CONNECTION_TYPE_ATSERVER);
   ctx->atserver_connection_started = true;
 
+  // add back hooks
+  ctx->atserver_connection.hooks = conn_hooks;
+
```

Comment:
originally tried implementing a reset function to replace with this, but it caused a slew of other problems. Simplest solutions is to save the address to the hooks struct and remove it from the ctx, that way it isn't overwritten by _free or _init. Then we simply write the address to the hooks back, this is also naturally resilient to the case where hooks are NULL.